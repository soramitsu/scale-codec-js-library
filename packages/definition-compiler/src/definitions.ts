export type NamespaceDefinition = Record<string, TypeDef>;

export type AliasDef = {
    ref: string;
};

export type ArrayDef = {
    item: string;
    len: number;
};

export type BytesArrayDef = {
    len: number;
};

export type VecDef = {
    item: string;
};

export type TupleDef = {
    items: string[];
};

export type StructDef = {
    fields: {
        name: string;
        ref: string;
    }[];
};

export type MapDef = {
    key: string;
    value: string;
};

export type SetDef = {
    entry: string;
};

export type EnumDef = {
    variants: EnumVariantDef[];
};

export type EnumVariantDef = {
    name: string;
    discriminant: number;
    ref?: string | null;
};

export type OptionDef = {
    some: string;
};

export type ResultDef = {
    ok: string;
    err: string;
};

export type WithTMark<T, M extends string> = T & {
    t: M;
};

export type TypeDef =
    | WithTMark<AliasDef, 'alias'>
    | WithTMark<ArrayDef, 'array'>
    | WithTMark<BytesArrayDef, 'bytes-array'>
    | WithTMark<VecDef, 'vec'>
    | WithTMark<TupleDef, 'tuple'>
    | WithTMark<StructDef, 'struct'>
    | WithTMark<MapDef, 'map'>
    | WithTMark<SetDef, 'set'>
    | WithTMark<EnumDef, 'enum'>
    | WithTMark<OptionDef, 'option'>
    | WithTMark<ResultDef, 'result'>;
