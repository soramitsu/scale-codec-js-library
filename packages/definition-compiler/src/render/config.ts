import { InjectionKey, inject, provide } from 'vue';

const key: InjectionKey<Config> = Symbol('render config');

export interface Config {
    importLib: string;
}

export function useConfig(): Config {
    const val = inject(key);
    if (!val) throw new Error('no config');
    return val;
}

export function provideConfig(conf: Config): void {
    provide(key, conf);
}
