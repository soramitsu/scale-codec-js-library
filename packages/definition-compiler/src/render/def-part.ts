export type DefPart = 'ty-decoded' | 'ty-encodable' | 'fn-decode' | 'fn-encode';

export const DefPartSuffixMap: { [K in DefPart]: string } = {
    'fn-decode': '_decode',
    'fn-encode': '_encode',
    'ty-decoded': '_Decoded',
    'ty-encodable': '_Encodable',
};

export function partToSuffix(val: DefPart) {
    return DefPartSuffixMap[val];
}
