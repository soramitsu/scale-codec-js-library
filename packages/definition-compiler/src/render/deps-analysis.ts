import { Set, Map, List } from 'immutable'
import Graph from 'tarjan-graph'

export type AdjacencyMap = Map<string, Set<string>>

/**
 * This function is an implementation of topological sort of a graph that is customized
 * to handle cycles
 */
export function optimizeDepsHierarchy(graph: AdjacencyMap): {
  /**
   * Topologically sorted graph vertices
   */
  sorted: List<string>
  /**
   * Map with data where cycles are detected
   */
  circuitsResolutions: AdjacencyMap
} {
  /**
   * 1. Find components
   * 2. Construct Map<string, Set<string>> with components (key - each vert, value - its component)
   * 3. Sort each component topologically
   * 4. Find circuit resolutions in each component (using sorted components from step 3)
   * 5. Sort initial graph alphabetically -> topologically
   * 6. Return result from 4 and 5
   */

  // 1.
  const components = findStronglyConnectedComponents(graph)

  // 2.
  const componentsMap = buildComponentsMap(components)

  // 3.
  const componentsSorted: Map<Set<string>, List<string>> = components.reduce(
    (map, x) => map.set(x, topologicalSortWithIgnoranceOfCircuits(x, graph)),
    Map(),
  )

  // 4.
  const circuitsResolutions: Map<string, Set<string>> = componentsSorted
    .valueSeq()
    .reduce((map, sortedComponent) => map.merge(findCircuitsResolutions(sortedComponent, graph)), Map())

  // 5.
  const verticesSorted = sortCyclicGraphTopologicallyUsingComponents(graph, componentsMap, componentsSorted)

  return {
    sorted: verticesSorted,
    circuitsResolutions,
  }
}

export function sortCyclicGraphTopologicallyUsingComponents(
  graph: AdjacencyMap,
  componentsMap: AdjacencyMap,
  componentsSorted: Map<Set<String>, List<string>>,
): List<string> {
  const used = Set<string>().asMutable()
  const stack = List<string>().asMutable()

  function dfs(v: string): void {
    const component = componentsMap.get(v)!
    used.merge(component) // itself with all siblings

    for (const u of graph.get(v)!) {
      if (!used.has(u)) dfs(u)
    }

    const sortedComponent = componentsSorted.get(component)!
    stack.concat(sortedComponent)
  }

  for (const v of graph.keySeq().sort()) {
    if (!used.has(v)) dfs(v)
  }

  return stack.asImmutable()
}

export function findCircuitsResolutions(sortedComponent: List<string>, graph: AdjacencyMap): Map<string, Set<string>> {
  const indices = sortedComponent.reduce<Map<string, number>>((map, v, i) => map.set(v, i), Map())

  return sortedComponent.reduce((map, v, i) => {
    return graph.get(v)!.reduce((map, u) => {
      const uIdx = indices.get(u)
      if (typeof uIdx !== 'number' || uIdx < i) return map
      return map.update(v, (set) => (set ?? Set()).add(u))
    }, map)
  }, Map())
}

export function graphToAdjacencyList(graph: AdjacencyMap): {
  adjacencyList: number[][]
  adjacencyListKeys: string[]
} {
  const keys: string[] = Array.from(graph.keySeq().sort().reverse())
  const keysIdx: Map<string, number> = keys.reduce((map, key, idx) => map.set(key, idx), Map<string, number>())
  const list: number[][] = new Array(keys.length)

  for (let i = 0; i < keys.length; i++) {
    list[i] = graph
      .get(keys[i])!
      .map((u) => keysIdx.get(u)!)
      .toArray()
  }

  console.log({ list, keys })

  return { adjacencyList: list, adjacencyListKeys: keys }
}

/**
 * Uses Tarjan's algorithm
 */
export function findStronglyConnectedComponents(graph: AdjacencyMap): Set<Set<string>> {
  const graphTarjan = new Graph()

  for (const [v, deps] of graph.entries()) {
    graphTarjan.add(v, [...deps])
  }

  const components = graphTarjan.getStronglyConnectedComponents()

  return components.reduce((set, arr) => set.add(Set(arr.map((x) => x.name))), Set<Set<string>>())
}

export function topologicalSortWithIgnoranceOfCircuits(component: Set<string>, graph: AdjacencyMap): List<string> {
  const used = Set<string>().asMutable()
  const stack = List<string>().asMutable()

  function dfs(v: string): void {
    used.add(v)
    for (const u of graph.get(v)!) {
      if (component.has(u) && !used.has(u)) dfs(u)
    }
    stack.push(v)
  }

  for (const v of component) {
    if (!used.has(v)) dfs(v)
  }

  return stack.asImmutable()
}

export function buildComponentsMap(components: Set<Set<string>>): Map<string, Set<string>> {
  return components.reduce((map, component) => component.reduce((map, v) => map.set(v, component), map), Map())
}
