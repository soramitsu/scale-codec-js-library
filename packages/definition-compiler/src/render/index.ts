import {
    NamespaceDefinition,
    RenderNamespaceDefinitionParams,
    TypeDef,
    DefEnumVariant,
    DefStructField,
} from '../types';
import { Set as SetImmutable } from 'immutable';
import { renderImports, createStateScope } from './util';
import { byValue, byString } from 'sort-es';
import { DefaultAvailableBuilders } from '../const';

function namespaceDefinitionToList(val: NamespaceDefinition): { tyName: string; def: TypeDef }[] {
    const items = Object.entries(val);
    items.sort(byValue((x) => x[0], byString()));
    return items.map(([tyName, def]) => ({ tyName, def }));
}

enum BaseType {
    Instance = 'ScaleInstance',
    Builder = 'ScaleBuilder',
    InstanceViaBuilder = 'InstanceViaBuilder',
    InnerValue = 'InnerValue',
    UnwrappedValue = 'UnwrappedValue',
    Enum = 'Enum',
    Valuable = 'Valuable',
    Option = 'Option',
    Result = 'Result',
}

// =========

interface RenderParams {
    runtimeLib: string;
    runtimeTypes: Set<String>;
    rollupSingleTuples: boolean;
}

const { within: withinRenderParams, use: useRenderParams } = createStateScope<RenderParams>();

// =========

interface ImportsCollector {
    collectRef: (ref: string) => void;
    collectImport: (name: string) => void;
    getRuntimeLibImports: () => Set<string>;
}

function instanceViaBuilder(ref: string): string {
    return `${touchBase(BaseType.InstanceViaBuilder)}<typeof ${touchRef(ref)}>`;
}

function createImportsCollector(): ImportsCollector {
    const { runtimeTypes } = useRenderParams();

    let imports = SetImmutable<string>();

    return {
        getRuntimeLibImports: () => new Set(imports),
        collectRef: (ref) => {
            if (runtimeTypes.has(ref)) {
                imports = imports.add(ref);
            }
        },
        collectImport: (name) => {
            imports = imports.add(name);
        },
    };
}

const { within: withinCollector, use: useCollector } = createStateScope<ImportsCollector>();

// =========

const { within: withinCurrentTyName, use: useCurrentTyName } = createStateScope<string>();

// =========

function renderBuilder(props: { valueTy: string | null; createHelper: string; createHelperArgs: string }): string {
    const ty = useCurrentTyName();
    const helperGeneric = props.valueTy ? `<${props.valueTy}>` : '';

    return `export var ${ty} = ${touchImport(props.createHelper)}${helperGeneric}('${ty}', ${props.createHelperArgs})`;
}

function touchRef(ref: string): string {
    useCollector().collectRef(ref);
    return ref;
}

function touchImport(name: string): string {
    useCollector().collectImport(name);
    return name;
}

function touchBase(ty: BaseType): string {
    useCollector().collectImport(ty);
    return ty;
}

/**
 * ref -> `() => ${ref}`
 */
function refFn(ref: string): string {
    return `() => ${touchRef(ref)}`;
}

function linesJoin(lines: string[], joiner = '\n\n'): string {
    return lines.join(joiner);
}

// =========

function renderAlias(to: string): string {
    return renderBuilder({
        valueTy: [BaseType.InnerValue, BaseType.UnwrappedValue]
            .map((x) => `${touchBase(x)}<typeof ${touchRef(to)}>`)
            .join(', '),
        createHelper: 'createAliasBuilder',
        createHelperArgs: refFn(to),
    });
}

function renderVoidAlias(): string {
    const { runtimeLib } = useRenderParams();

    return renderImport({ nameInModule: 'Void', module: runtimeLib });
}

function renderVec(item: string): string {
    return renderBuilder({
        valueTy: `${instanceViaBuilder(item)}[]`,
        createHelper: 'createVecBuilder',
        createHelperArgs: refFn(item),
    });
}

function renderStruct(fields: DefStructField[]): string {
    if (!fields.length) {
        return renderVoidAlias();
    }

    const valueTypeFields: string[] = fields.map((x) => `${x.name}: ${instanceViaBuilder(x.ref)}`);

    const schemaItems = fields.map((x) => `['${x.name}', ${refFn(x.ref)}]`);
    const schema = `[${schemaItems.join(', ')}]`;

    return renderBuilder({
        valueTy: `{\n    ${valueTypeFields.join(',\n    ')}\n}`,
        createHelper: 'createStructBuilder',
        createHelperArgs: `${schema}`,
    });
}

function renderTuple(refs: string[]): string {
    if (!refs.length) {
        return renderVoidAlias();
    }

    const { rollupSingleTuples } = useRenderParams();
    if (rollupSingleTuples && refs.length === 1) return renderAlias(refs[0]);

    const valueEntries: string[] = refs.map(instanceViaBuilder);
    const codecs: string[] = refs.map(refFn);

    return renderBuilder({
        valueTy: `[\n    ${valueEntries.join(',\n    ')}\n]`,
        createHelper: 'createTupleBuilder',
        createHelperArgs: `[${codecs.join(', ')}]`,
    });
}

function renderEnum(variants: DefEnumVariant[]): string {
    const definitionTyLines: string[] = variants.map((x) => {
        const right = x.ref ? `${touchBase(BaseType.Valuable)}<${instanceViaBuilder(x.ref)}>` : 'null';
        return `${x.name}: ${right}`;
    });

    const schemaLines: string[] = variants.map((x) => {
        const items = [x.discriminant, `'${x.name}'`];
        x.ref && items.push(refFn(x.ref));
        return `[${items.join(', ')}]`;
    });

    return renderBuilder({
        valueTy: `${touchBase(BaseType.Enum)}<{\n    ${definitionTyLines.join(',\n    ')}\n}>`,
        createHelper: 'createEnumBuilder',
        createHelperArgs: `[${schemaLines.join(', ')}]`,
    });
}

function renderSet(item: string): string {
    return renderBuilder({
        valueTy: `Set<${instanceViaBuilder(item)}>`,
        createHelper: 'createSetBuilder',
        createHelperArgs: refFn(item),
    });
}

function renderMap(key: string, value: string): string {
    return renderBuilder({
        valueTy: `Map<${instanceViaBuilder(key)}, ${instanceViaBuilder(value)}>`,
        createHelper: 'createMapBuilder',
        createHelperArgs: [key, value].map(refFn).join(', '),
    });
}

function renderArray(item: string, len: number): string {
    return renderBuilder({
        valueTy: `${instanceViaBuilder(item)}[]`,
        createHelper: `createArrayBuilder`,
        createHelperArgs: `${refFn(item)}, ${len}`,
    });
}

function renderBytesArray(len: number): string {
    return renderBuilder({
        valueTy: null,
        createHelper: 'createBytesArrayBuilder',
        createHelperArgs: `${len}`,
    });
}

function renderOption(some: string): string {
    return renderBuilder({
        valueTy: `${touchBase(BaseType.Option)}<${instanceViaBuilder(some)}>`,
        createHelper: 'createOptionBuilder',
        createHelperArgs: refFn(some),
    });
}

function renderResult(ok: string, err: string): string {
    return renderBuilder({
        valueTy: `${touchBase(BaseType.Result)}<${instanceViaBuilder(ok)}, ${instanceViaBuilder(err)}>`,
        createHelper: 'createResultBuilder',
        createHelperArgs: [ok, err].map(refFn).join(', '),
    });
}

function renderImport({ nameInModule, module: moduleName }: { nameInModule?: string | null; module: string }): string {
    const ty = useCurrentTyName();

    return linesJoin(
        [renderImports([nameInModule ? { source: nameInModule, as: ty } : ty], moduleName), `export { ${ty} }`],
        '\n',
    );
}

function renderParticularDef(tyName: string, def: TypeDef): string {
    const particularRendered: string = withinCurrentTyName(tyName, () => {
        switch (def.t) {
            case 'alias':
                return renderAlias(def.ref);
            case 'vec':
                return renderVec(def.item);
            case 'struct':
                return renderStruct(def.fields);
            case 'tuple':
                return renderTuple(def.items);
            case 'enum':
                return renderEnum(def.variants);
            case 'set':
                return renderSet(def.entry);
            case 'map':
                return renderMap(def.key, def.value);
            case 'array':
                return renderArray(def.item, def.len);
            case 'bytes-array':
                return renderBytesArray(def.len);
            case 'option':
                return renderOption(def.some);
            case 'result':
                return renderResult(def.ok, def.err);
            case 'import':
                return renderImport(def);
            default: {
                const uncovered: never = def;
                throw new Error(`Undefined type definition: ${uncovered}`);
            }
        }
    });

    return particularRendered;
}

function renderPreamble(): string {
    const { runtimeLib } = useRenderParams();
    const { getRuntimeLibImports: getCoreImports } = useCollector();

    const lines = [];

    const imports = getCoreImports();
    if (imports.size) {
        lines.push(renderImports(imports, runtimeLib));
    }

    return linesJoin(lines);
}

/**
 * Renders provided definition into a valid TypeScript code.
 */
export function renderNamespaceDefinition(
    definition: NamespaceDefinition,
    params?: RenderNamespaceDefinitionParams,
): string {
    return withinRenderParams(
        {
            runtimeLib: params?.runtimeLib ?? '@scale-codec/definition-runtime',
            runtimeTypes: params?.runtimeTypes ?? DefaultAvailableBuilders,
            rollupSingleTuples: params?.rollupSingleTuplesIntoAliases ?? false,
        },
        () =>
            withinCollector(createImportsCollector(), () => {
                const renderedTypes = namespaceDefinitionToList(definition)
                    .map(({ tyName, def }) => renderParticularDef(tyName, def))
                    .join('\n\n');

                const preamble = renderPreamble();

                return [preamble, renderedTypes].join('\n\n').trim() + '\n';
            }),
    );
}
