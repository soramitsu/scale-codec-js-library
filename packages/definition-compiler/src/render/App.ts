import { defineComponent, PropType, computed, compile } from 'vue';
import { NamespaceDefinition, TypeDef } from '../definitions';
import { byValue, byString } from 'sort-es';
import { provideConfig } from './config';

import Collector from './components/Collector';
import Preamble from './components/Preamble';
import DefRenderer from './components/DefRenderer';

export default defineComponent({
    components: {
        Collector,
        Preamble,
        DefRenderer,
    },
    props: {
        defmap: {
            type: Object as PropType<NamespaceDefinition>,
            required: true,
        },
        importLib: {
            type: String,
            default: 'sample',
        },
    },
    setup(props) {
        const defsList = computed<{ tyName: String; def: TypeDef }[]>(() => {
            const items = Object.entries(props.defmap);
            items.sort(byValue((x) => x[0], byString()));
            return items.map(([tyName, def]) => ({ tyName, def }));
        });

        provideConfig({
            importLib: props.importLib,
        });

        return {
            defsList,
        };
    },
    render: compile(`
        <collector>
            <preamble />


            <template v-for="x in defsList" :key="x.tyName">
                <w t="\n\n" />
                <def-renderer v-bind="x" />
            </template>
        </collector>
    `),
});
