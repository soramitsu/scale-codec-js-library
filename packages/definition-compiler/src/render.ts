import { createApp, renderApp } from './vue-code';
import { NamespaceDefinition } from './definitions';
import App from './render/App';
import { WithDefPart, AddPartSuffix } from './render/components/current-def-part';
import { WithCurrentTypeName, TyName } from './render/components/current-type-name';
import Whitespace from './render/components/Whitespace';
import { TypeReference, Core } from './render/components/links';
import TypeExport from './render/components/TypeExport';
import Fields from './render/components/Fields';

export async function renderNamespaceDefinition(
    def: NamespaceDefinition,
    params: {
        importLib: string;
    },
): Promise<string> {
    const app = createApp(App, { defmap: def, importLib: params.importLib })
        .component('WithDefPart', WithDefPart)
        .component('WithCurrentTypeName', WithCurrentTypeName)
        .component('TyName', TyName)
        .component('Ref', TypeReference)
        .component('Core', Core)
        .component('W', Whitespace)
        .component('Fields', Fields)
        .component('AddPartSuffix', AddPartSuffix)
        .component('Export', TypeExport);

    return renderApp(app);
}
