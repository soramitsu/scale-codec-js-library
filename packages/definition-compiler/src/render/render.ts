import { createApp, renderApp } from './vue-code';
import { NamespaceDefinition } from '../definitions';
import App from './App';
import { WithDefPart, AddPartSuffix } from './components/current-def-part';
import { WithCurrentTypeName, TyName } from './components/current-type-name';
import Whitespace from './components/Whitespace';
import { TypeReference, Core } from './components/links';
import TypeExport from './components/TypeExport';
import Fields from './components/Fields';
import UseConfig from './components/UseConfig';
import VoidAlias from './components/VoidAlias';

export interface RenderNamespaceDefinitionParams {
    /**
     * Runtime library with STD codecs + reexports from the core library.
     *
     * @remarks
     * See `@scale-codec/definition-runtime` library - it exists specially for it.
     */
    importLib: string;
    /**
     * Single tuples are always an arrays with a single element. It is possible to make final code cleaner
     * (and a bit performant) if render such tuples just as aliases for the inner element. It is optional feature.
     */
    rollupSingleTuplesIntoAliases?: boolean;
}

/**
 * Renders provided definition into a valid TypeScript code.
 */
export async function renderNamespaceDefinition(
    def: NamespaceDefinition,
    params: RenderNamespaceDefinitionParams,
): Promise<string> {
    const app = createApp(App, {
        defmap: def,
        importLib: params.importLib,
        rollupSingleTuples: params.rollupSingleTuplesIntoAliases ?? false,
    })
        .component('WithDefPart', WithDefPart)
        .component('WithCurrentTypeName', WithCurrentTypeName)
        .component('TyName', TyName)
        .component('Ref', TypeReference)
        .component('Core', Core)
        .component('W', Whitespace)
        .component('Fields', Fields)
        .component('AddPartSuffix', AddPartSuffix)
        .component('UseConfig', UseConfig)
        .component('VoidAlias', VoidAlias)
        .component('Export', TypeExport);

    return renderApp(app);
}
