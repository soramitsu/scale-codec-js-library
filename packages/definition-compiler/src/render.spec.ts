import { NamespaceDefinition } from './definitions';
import { renderNamespaceDefinition } from './render';

async function expectRenderToMatchSnapshot(
    def: NamespaceDefinition,
    importLib = '@scale-codec/definition-compiler-runtime',
) {
    expect(await renderNamespaceDefinition(def, { importLib })).toMatchSnapshot();
}

it('Renders vec', () => {
    return expectRenderToMatchSnapshot({
        Vec_str: {
            t: 'vec',
            item: 'str',
        },
    });
});

it('Renders struct + tuple', () => {
    return expectRenderToMatchSnapshot({
        Structural: {
            t: 'struct',
            fields: [
                {
                    name: 'nums',
                    ref: 'Tuple_u8_i128',
                },
                {
                    name: 'mur',
                    ref: 'bool',
                },
            ],
        },
        Tuple_u8_i128: {
            t: 'tuple',
            items: ['u8', 'i128'],
        },
    });
});

it('Renders enum', () => {
    return expectRenderToMatchSnapshot({
        Message: {
            t: 'enum',
            variants: [
                {
                    name: 'Quit',
                    discriminant: 0,
                },
                {
                    name: 'Greeting',
                    discriminant: 1,
                    ref: 'str',
                },
            ],
        },
    });
});

it('Renders set', () => {
    return expectRenderToMatchSnapshot({
        Set_Message: {
            t: 'set',
            entry: 'Message',
        },
    });
});
