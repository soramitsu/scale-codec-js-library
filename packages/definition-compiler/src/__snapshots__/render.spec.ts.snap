// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders array 1`] = `
"import {
  EncodeSkippable,
  decodeArray,
  encodeArray,
  wrapSkippableEncode,
} from '@scale-codec/definition-compiler-runtime'

// Array_Item_15

export type Array_Item_15_Decoded = Item_Decoded[]

export type Array_Item_15_Encodable = (Item_Encodable | EncodeSkippable)[]

const Array_Item_15_item_encode = wrapSkippableEncode(Item_encode)
const Array_Item_15_len = 15

export function Array_Item_15_decode(bytes: Uint8Array): Array_Item_15_Decoded {
  return decodeArray(bytes, Item_decode, Array_Item_15_len)
}

export function Array_Item_15_encode(encodable: Array_Item_15_Encodable): Uint8Array {
  return encodeArray(encodable, Array_Item_15_item_encode, Array_Item_15_len)
}
"
`;

exports[`Renders bytes array 1`] = `
"import { decodeBytesArray, encodeBytesArray } from '@scale-codec/definition-compiler-runtime'

// Hash

export type Hash_Decoded = Uint8Array

export type Hash_Encodable = Uint8Array

const Hash_len = 64

export function Hash_decode(bytes: Uint8Array): Hash_Decoded {
  return decodeBytesArray(bytes, Hash_len)
}

export function Hash_encode(encodable: Hash_Encodable): Uint8Array {
  return encodeBytesArray(encodable, Hash_len)
}
"
`;

exports[`Renders enum 1`] = `
"import {
  EncodeSkippable,
  Enum,
  Valuable,
  decodeEnum,
  encodeEnum,
  str_Decoded,
  str_Encodable,
  str_decode,
  str_encode,
} from '@scale-codec/definition-compiler-runtime'

// Message

export type Message_Decoded = Enum<{
  Quit: null
  Greeting: Valuable<str_Decoded>
}>

export type Message_Encodable = Enum<{
  Quit: null
  Greeting: Valuable<str_Encodable | EncodeSkippable>
}>

// Message enum tools

const Message_decoders = {
  0: { v: 'Quit' },
  1: { v: 'Greeting', decode: str_decode },
}
const Message_encoders = {
  Quit: { d: 0 },
  Greeting: { d: 1, encode: str_encode },
}

// Message tools end

export function Message_decode(bytes: Uint8Array): Message_Decoded {
  return decodeEnum(bytes, Message_decoders)
}

export function Message_encode(encodable: Message_Encodable): Uint8Array {
  return encodeEnum(encodable, Message_encoders)
}
"
`;

exports[`Renders map 1`] = `
"import {
  EncodeSkippable,
  decodeMap,
  encodeMap,
  str_Decoded,
  str_Encodable,
  str_decode,
  str_encode,
  wrapSkippableEncode,
} from '@scale-codec/definition-compiler-runtime'

// Map_str_Message

export type Map_str_Message_Decoded = Map<str_Decoded, Message_Decoded>

export type Map_str_Message_Encodable = Map<str_Encodable | EncodeSkippable, Message_Encodable | EncodeSkippable>

// Map_str_Message map tools

const [Map_str_Message_encode_key, Map_str_Message_encode_value] = [str_encode, Message_encode].map(wrapSkippableEncode)

// Map_str_Message tools end

export function Map_str_Message_decode(bytes: Uint8Array): Map_str_Message_Decoded {
  return decodeMap(bytes, str_decode, Message_decode)
}

export function Map_str_Message_encode(encodable: Map_str_Message_Encodable): Uint8Array {
  return encodeMap(encodable, Map_str_Message_encode_key, Map_str_Message_encode_value)
}
"
`;

exports[`Renders set 1`] = `
"import { EncodeSkippable, decodeSet, encodeSet, wrapSkippableEncode } from '@scale-codec/definition-compiler-runtime'

// Set_Message

export type Set_Message_Decoded = Set<Message_Decoded>

export type Set_Message_Encodable = Set<Message_Encodable | EncodeSkippable>

// Set_Message set tools

const Set_Message_entry_encode = wrapSkippableEncode(Message_encode)

// Set_Message tools end

export function Set_Message_decode(bytes: Uint8Array): Set_Message_Decoded {
  return decodeSet(bytes, Message_decode)
}

export function Set_Message_encode(encodable: Set_Message_Encodable): Uint8Array {
  return encodeSet(encodable, Set_Message_entry_encode)
}
"
`;

exports[`Renders struct + tuple 1`] = `
"import {
  EncodeSkippable,
  bool_Decoded,
  bool_Encodable,
  bool_decode,
  bool_encode,
  decodeStruct,
  decodeTuple,
  encodeStruct,
  encodeTuple,
  i128_Decoded,
  i128_Encodable,
  i128_decode,
  i128_encode,
  u8_Decoded,
  u8_Encodable,
  u8_decode,
  u8_encode,
  wrapSkippableEncode,
} from '@scale-codec/definition-compiler-runtime'

// Structural

export type Structural_Decoded = {
  nums: Tuple_u8_i128_Decoded
  mur: bool_Decoded
}

export type Structural_Encodable = {
  nums: Tuple_u8_i128_Encodable | EncodeSkippable
  mur: bool_Encodable | EncodeSkippable
}

// Structural struct tools

const Structural_order = ['nums', 'mur']
const Structural_decoders = {
  nums: Tuple_u8_i128_decode,
  mur: bool_decode,
}
const Structural_encoders = {
  nums: wrapSkippableEncode(Tuple_u8_i128_encode),
  mur: wrapSkippableEncode(bool_encode),
}

// Structural tools end

export function Structural_decode(bytes: Uint8Array): Structural_Decoded {
  return decodeStruct(bytes, Structural_decoders, Structural_order)
}

export function Structural_encode(encodable: Structural_Encodable): Uint8Array {
  return encodeStruct(encodable, Structural_encoders, Structural_order)
}

// Tuple_u8_i128

export type Tuple_u8_i128_Decoded = [u8_Decoded, i128_Decoded]

export type Tuple_u8_i128_Encodable = [u8_Encodable | EncodeSkippable, i128_Encodable | EncodeSkippable]

// Tuple_u8_i128 tuple-related tools

const Tuple_u8_i128_decoders = [u8_decode, i128_decode]
const Tuple_u8_i128_encoders = [u8_encode, i128_encode].wrap(wrapSkippableEncode)

// Tuple_u8_i128 tools end

export function Tuple_u8_i128_decode(bytes: Uint8Array): Tuple_u8_i128_Decoded {
  return decodeTuple(bytes, Tuple_u8_i128_decoders)
}

export function Tuple_u8_i128_encode(encodable: Tuple_u8_i128_Encodable): Uint8Array {
  return encodeTuple(encodable, Tuple_u8_i128_encoders)
}
"
`;

exports[`Renders vec 1`] = `
"import {
  EncodeSkippable,
  decodeVec,
  encodeVec,
  str_Decoded,
  str_Encodable,
  str_decode,
  str_encode,
} from '@scale-codec/definition-compiler-runtime'

// Vec_str

export type Vec_str_Decoded = str_Decoded[]

export type Vec_str_Encodable = (str_Encodable | EncodeSkippable)[]

export function Vec_str_decode(bytes: Uint8Array): Vec_str_Decoded {
  return decodeVec(bytes, str_decode)
}

export function Vec_str_encode(encodable: Vec_str_Encodable): Uint8Array {
  return encodeVec(encodable, str_encode)
}
"
`;
